File: trainer-flutter/lib/models/training_plan.dart
import 'dart:convert';

class TrainingPlanRow {
  String exercise;
  String device;
  String position;
  String weight;
  String sets;        // e.g. "3×12"
  List<String> dates;
  bool liked;
  bool disliked;
  List<int> timers;   // saved timer presets in seconds (e.g. [30, 60, 120])
  String comment;     // per-exercise comment / note (legacy, now server-side)

  TrainingPlanRow({
    this.exercise = '',
    this.device = '',
    this.position = '',
    this.weight = '',
    this.sets = '',
    List<String>? dates,
    this.liked = false,
    this.disliked = false,
    List<int>? timers,
    this.comment = '',
  }) : dates = dates ?? List.filled(8, ''),
       timers = timers ?? [];

  factory TrainingPlanRow.fromJson(Map<String, dynamic> json) {
    return TrainingPlanRow(
      exercise: json['exercise']?.toString() ?? '',
      device: json['device']?.toString() ?? '',
      position: json['position']?.toString() ?? '',
      weight: json['weight']?.toString() ?? '',
      sets: json['sets']?.toString() ?? '',
      dates: json['dates'] is List
          ? List<String>.from((json['dates'] as List).map((e) => e?.toString() ?? ''))
          : List.filled(8, ''),
      liked: json['liked'] == true || json['liked'] == 1 || json['liked'] == '1',
      disliked: json['disliked'] == true || json['disliked'] == 1 || json['disliked'] == '1',
      timers: _parseTimers(json),
      comment: json['comment']?.toString() ?? '',
    );
  }

  static int _parseInt(dynamic v) {
    if (v == null) return 0;
    if (v is int) return v;
    if (v is String) return int.tryParse(v) ?? 0;
    return 0;
  }

  /// Parse timers from JSON — backwards-compatible with old `timer` (int) field
  static List<int> _parseTimers(Map<String, dynamic> json) {
    // New format: timers as list
    if (json['timers'] is List) {
      return List<int>.from(
        (json['timers'] as List).map((e) => _parseInt(e)).where((v) => v > 0),
      );
    }
    // Legacy format: single timer int
    final single = _parseInt(json['timer']);
    return single > 0 ? [single] : [];
  }

  Map<String, dynamic> toJson() => {
        'exercise': exercise,
        'device': device,
        'position': position,
        'weight': weight,
        if (sets.isNotEmpty) 'sets': sets,
        'dates': dates,
        if (liked) 'liked': true,
        if (disliked) 'disliked': true,
        if (timers.isNotEmpty) 'timers': timers,
        if (comment.isNotEmpty) 'comment': comment,
      };
}

class TrainingPlanValues {
  List<TrainingPlanRow> sonsomo;
  List<TrainingPlanRow> main;
  List<TrainingPlanRow> core;
  List<TrainingPlanRow> mobility;
  List<String> dates;

  TrainingPlanValues({
    List<TrainingPlanRow>? sonsomo,
    List<TrainingPlanRow>? main,
    List<TrainingPlanRow>? core,
    List<TrainingPlanRow>? mobility,
    List<String>? dates,
  })  : sonsomo = sonsomo ?? [],
        main = main ?? [],
        core = core ?? [],
        mobility = mobility ?? [],
        dates = dates ?? List.filled(8, '');

  factory TrainingPlanValues.fromJson(Map<String, dynamic> json) {
    List<TrainingPlanRow> parseRows(dynamic list) {
      if (list is List) {
        return list
            .map((e) => TrainingPlanRow.fromJson(e as Map<String, dynamic>))
            .toList();
      }
      return [];
    }

    return TrainingPlanValues(
      sonsomo: parseRows(json['sonsomo']),
      main: parseRows(json['main']),
      core: parseRows(json['core']),
      mobility: parseRows(json['mobility']),
      dates: json['dates'] is List
          ? List<String>.from((json['dates'] as List).map((e) => e?.toString() ?? ''))
          : List.filled(8, ''),
    );
  }

  Map<String, dynamic> toJson() => {
        'sonsomo': sonsomo.map((r) => r.toJson()).toList(),
        'main': main.map((r) => r.toJson()).toList(),
        'core': core.map((r) => r.toJson()).toList(),
        'mobility': mobility.map((r) => r.toJson()).toList(),
        'dates': dates,
      };
}

class TrainingPlan {
  final int? id;
  final int? clientId;
  String? name;
  TrainingPlanValues values;
  String? createdAt;

  TrainingPlan({
    this.id,
    this.clientId,
    this.name,
    TrainingPlanValues? values,
    this.createdAt,
  }) : values = values ?? TrainingPlanValues();

  factory TrainingPlan.fromJson(Map<String, dynamic> json) {
    TrainingPlanValues parsedValues;
    final rawValues = json['values'];
    if (rawValues is String && rawValues.isNotEmpty) {
      try {
        final decoded = jsonDecode(rawValues) as Map<String, dynamic>;
        parsedValues = TrainingPlanValues.fromJson(decoded);
      } catch (_) {
        parsedValues = TrainingPlanValues();
      }
    } else if (rawValues is Map<String, dynamic>) {
      parsedValues = TrainingPlanValues.fromJson(rawValues);
    } else {
      parsedValues = TrainingPlanValues();
    }

    return TrainingPlan(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id']?.toString() ?? ''),
      clientId: json['client_id'] is int
          ? json['client_id']
          : int.tryParse(json['client_id']?.toString() ?? ''),
      name: json['name']?.toString(),
      values: parsedValues,
      createdAt: json['created_at']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        if (clientId != null) 'client_id': clientId,
        if (name != null) 'name': name,
        'values': jsonEncode(values.toJson()),
      };
}

File: trainer-ios/SihlTrainer/Models/TrainingPlan.swift
import Foundation

/// Eine Übungszeile im Trainingsplan.
/// Pendant zu `TrainingPlanRow` aus `models/training_plan.dart`.
///
/// `id` ist rein lokal (SwiftUI braucht eine stabile Identität beim
/// Bearbeiten) und wird nicht ans Backend geschickt.
struct TrainingPlanRow: Identifiable, Equatable {
    let id = UUID()
    var exercise = ""
    var device = ""
    var position = ""
    var weight = ""
    var sets = ""            // z.B. "3×12"
    var comment = ""
    var liked = false
    var disliked = false
    var timers: [Int] = []   // gespeicherte Timer-Vorgaben in Sekunden

    init() {}

    init(json: [String: Any]) {
        exercise = JSON.string(json, "exercise") ?? ""
        device = JSON.string(json, "device") ?? ""
        position = JSON.string(json, "position") ?? ""
        weight = JSON.string(json, "weight") ?? ""
        sets = JSON.string(json, "sets") ?? ""
        comment = JSON.string(json, "comment") ?? ""
        liked = JSON.bool(json, "liked")
        disliked = JSON.bool(json, "disliked")
        timers = TrainingPlanRow.parseTimers(json)
    }

    /// Neues Format: `timers` als Liste. Altes Format: ein einzelnes `timer`.
    private static func parseTimers(_ json: [String: Any]) -> [Int] {
        if let list = json["timers"] as? [Any] {
            return list.compactMap { value in
                if let number = value as? Int { return number > 0 ? number : nil }
                if let text = value as? String, let number = Int(text) { return number > 0 ? number : nil }
                return nil
            }
        }
        if let single = JSON.int(json, "timer"), single > 0 { return [single] }
        return []
    }

    /// Nur gesetzte Felder senden — wie `toJson()` in Dart.
    var json: [String: Any] {
        var result: [String: Any] = [
            "exercise": exercise,
            "device": device,
            "position": position,
            "weight": weight,
        ]
        if !sets.isEmpty { result["sets"] = sets }
        if liked { result["liked"] = true }
        if disliked { result["disliked"] = true }
        if !timers.isEmpty { result["timers"] = timers }
        if !comment.isEmpty { result["comment"] = comment }
        return result
    }

    var isEmpty: Bool {
        exercise.isEmpty && device.isEmpty && position.isEmpty && weight.isEmpty && sets.isEmpty
    }
}

/// Die vier Abschnitte eines Plans.
enum PlanSection: String, CaseIterable, Identifiable {
    case sonsomo, main, core, mobility

    var id: String { rawValue }

    /// Anzeigenamen wie in `training_plan_detail_screen.dart`.
    var title: String {
        switch self {
        case .sonsomo:  return "AUFWÄRMEN"
        case .main:     return "HAUPTTRAINING"
        case .core:     return "CORE"
        case .mobility: return "MOBILITÄT"
        }
    }

    var icon: String {
        switch self {
        case .sonsomo:  return "figure.cooldown"
        case .main:     return "dumbbell"
        case .core:     return "figure.core.training"
        case .mobility: return "figure.flexibility"
        }
    }
}

/// Inhalt eines Plans: vier Abschnitte plus acht Termin-Spalten.
struct TrainingPlanValues: Equatable {
    var sections: [PlanSection: [TrainingPlanRow]]
    var dates: [String]

    init() {
        sections = Dictionary(uniqueKeysWithValues: PlanSection.allCases.map { ($0, []) })
        dates = Array(repeating: "", count: 8)
    }

    init(json: [String: Any]) {
        self.init()
        for section in PlanSection.allCases {
            let list = json[section.rawValue] as? [[String: Any]] ?? []
            sections[section] = list.map(TrainingPlanRow.init(json:))
        }
        if let raw = json["dates"] as? [Any] {
            var parsed = raw.map { String(describing: $0 is NSNull ? "" : $0) }
            while parsed.count < 8 { parsed.append("") }
            dates = Array(parsed.prefix(8))
        }
    }

    var json: [String: Any] {
        var result: [String: Any] = ["dates": dates]
        for section in PlanSection.allCases {
            result[section.rawValue] = (sections[section] ?? []).map(\.json)
        }
        return result
    }

    var totalRows: Int {
        PlanSection.allCases.reduce(0) { $0 + (sections[$1]?.count ?? 0) }
    }

    subscript(section: PlanSection) -> [TrainingPlanRow] {
        get { sections[section] ?? [] }
        set { sections[section] = newValue }
    }
}

/// Ein Trainingsplan. Pendant zu `TrainingPlan` aus `models/training_plan.dart`.
struct TrainingPlan: Identifiable, Equatable {
    var id: Int?
    var clientId: Int?
    var name: String?
    var values: TrainingPlanValues
    var createdAt: String?
    /// 'draft' = Trainer bearbeitet noch, 'published' = für den Kunden freigegeben.
    var status: String?
    var publishedAt: String?

    var isPublished: Bool { status == "published" }

    init(clientId: Int?, name: String? = nil) {
        self.clientId = clientId
        self.name = name
        values = TrainingPlanValues()
        status = "draft"
    }

    init(json: [String: Any]) {
        id = JSON.int(json, "id")
        clientId = JSON.int(json, "client_id", "clientId")
        name = JSON.string(json, "name")
        createdAt = JSON.string(json, "created_at", "createdAt")
        status = JSON.string(json, "status")
        publishedAt = JSON.string(json, "publishedAt", "published_at")

        // `values` kommt je nach Backend als JSON-String oder als Objekt.
        if let text = json["values"] as? String, !text.isEmpty,
           let data = text.data(using: .utf8),
           let decoded = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            values = TrainingPlanValues(json: decoded)
        } else if let object = json["values"] as? [String: Any] {
            values = TrainingPlanValues(json: object)
        } else {
            values = TrainingPlanValues()
        }
    }

    /// Speicher-Payload. `values` geht als JSON-String raus — so erwartet es
    /// das Backend (Legacy-Spalte).
    var savePayload: [String: Any] {
        var result: [String: Any] = [:]
        if let id { result["id"] = id }
        if let clientId { result["client_id"] = clientId }
        if let name { result["name"] = name }
        if let data = try? JSONSerialization.data(withJSONObject: values.json),
           let text = String(data: data, encoding: .utf8) {
            result["values"] = text
        }
        return result
    }
}
